iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/compaction/IcebergMajorQueryCompactor.java

```diff
@@ -44,22 +55,71 @@ public boolean run(CompactorContext context) throws IOException, HiveException,
+
+    if (partSpec == null) {
+      HiveConf.setVar(conf, ConfVars.REWRITE_POLICY, RewritePolicy.ALL_PARTITIONS.name());
+      compactionQuery = String.format("insert overwrite table %s select * from %<s", compactTableName);
+    } else {
+      Table table = IcebergTableUtil.getTable(conf, context.getTable());
+      List<String> partFields = table.spec().fields().stream().map(PartitionField::name).collect(Collectors.toList());
+      String queryFields = table.schema().columns().stream().map(Types.NestedField::name)
+          .filter(f -> !partFields.contains(f)).collect(Collectors.joining(","));
+      PartitionData partitionData = DataFiles.data(table.spec(), partSpec);
+      compactionQuery = String.format("insert overwrite table %1$s partition(%2$s) select %4$s from %1$s where %3$s",
+          compactTableName, partDataToSQL(partitionData, partSpec, ","),
+          partDataToSQL(partitionData, partSpec, " and "), queryFields);
+    }
+
+    SessionState sessionState = setupQueryCompactionSession(conf, context.getCompactionInfo(), tblProperties);
 
-    SessionState sessionState = setupQueryCompactionSession(context.getConf(),
-        context.getCompactionInfo(), tblProperties);
-    HiveConf.setVar(context.getConf(), ConfVars.REWRITE_POLICY, RewritePolicy.ALL_PARTITIONS.name());
     try {
-      DriverUtils.runOnDriver(context.getConf(), sessionState, compactionQuery);
+      DriverUtils.runOnDriver(conf, sessionState, compactionQuery);
       LOG.info("Completed compaction for table {}", compactTableName);
+      return true;
     } catch (HiveException e) {
-      LOG.error("Error doing query based {} compaction", RewritePolicy.ALL_PARTITIONS.name(), e);
-      throw new RuntimeException(e);
+      LOG.error("Failed compacting table {}", compactTableName, e);
+      throw e;
     } finally {
       sessionState.setCompaction(false);
     }
+  }
+
+  private String partDataToSQL(PartitionData partitionData, String partSpec, String delimiter)
+      throws HiveException {
+
+    try {
+      Map<String, String> partSpecMap = Warehouse.makeSpecFromName(partSpec);
+      StringBuilder sb = new StringBuilder();
+
+      for (int i = 0; i < partitionData.size(); ++i) {
+        if (i > 0) {
+          sb.append(delimiter);
+        }
+
+        String quoteOpt = "";
+        if (partitionData.getType(i).typeId() == Type.TypeID.STRING ||
+            partitionData.getType(i).typeId() == Type.TypeID.DATE ||
+            partitionData.getType(i).typeId() == Type.TypeID.TIME ||
+            partitionData.getType(i).typeId() == Type.TypeID.TIMESTAMP ||
+            partitionData.getType(i).typeId() == Type.TypeID.BINARY) {
+          quoteOpt = "'";
+        }
 
-    return true;
+        String fieldName = partitionData.getSchema().getFields().get(i).name();
+
+        sb.append(fieldName)
+            .append("=")
+            .append(quoteOpt)
+            .append(partSpecMap.get(fieldName))
+             .append(quoteOpt);
```

Comment:
nit: remove extra space
also, you could simplify the code with 
````
HiveUtils.unparseIdentifier(fieldName)
  .append("=")  
  .append(TypeInfoUtils.convertStringToLiteralForSQL(
    partSpecMap.get(fieldName), 
    ((PrimitiveTypeInfo) TypeInfoUtils.getTypeInfoFromTypeString(partitionData.getType(i).typeId())).getPrimitiveCategory() 
````
check `DDLPlanUtils.getPartitionActualName`

Reply:
`getTypeInfoFromTypeString()` didn't work with getting the type from partitionData, but I made it work a little differently just like in `TruncateTableAnalyzer.constructDeleteQuery()`.

Reply:
how about 
````
PrimitiveObjectInspector primitiveObjInspector = IcebergObjectInspector.primitive(partitionData.getType(i))
TypeInfoUtils.convertStringToLiteralForSQL(value, primitiveObjInspector.getPrimitiveCategory())
````

Reply:
Thanks, done.